### PR TITLE
Support for Nokogiri Fragment in all adapters

### PR DESCRIPTION
**Pull request by lnhrdt**
based on the work in https://github.com/premailer/premailer/pull/348

### CODE REVIEW

**Comment by akzhan** [comment]
Great, thanks

**Comment by akzhan** [comment]
New release will be until weekend.

**Comment by lnhrdt** [comment]
@akzhan thank you. I'll plan accordingly. Let me know if there's anything I can help with!

File: lib/premailer/adapter/nokogiri_fast.rb [hunk 1]
```diff
@@ -150,12 +150,19 @@ def to_inline_css
       # @return [::Nokogiri::XML] a document.
       def write_unmergable_css_rules(doc, unmergable_rules) # :nodoc:
         styles = unmergable_rules.to_s
-        return doc  if styles.empty?
-        style_tag = doc.create_element "style", styles
-        head = doc.at_css('head')
-        head ||=  doc.root.first_element_child.add_previous_sibling(doc.create_element "head")  if doc.root && doc.root.first_element_child
-        head ||=  doc.add_child(doc.create_element "head")
-        head << style_tag
+        unless styles.empty?
```
Comment (grosser) [inline lib/premailer/adapter/nokogiri_fast.rb]: FYI unless empty -> if any

File: lib/premailer/adapter/nokogumbo.rb [hunk 2]
```diff
@@ -148,12 +148,19 @@ def to_inline_css
...
+            style_tag.content = styles
+            doc.add_child(style_tag)
+          else
+            style_tag = doc.create_element "style", styles
+            head = doc.at_css('head')
+            head ||= doc.root.first_element_child.add_previous_sibling(doc.create_element "head") if doc.root && doc.root.first_element_child
+            head ||= doc.add_child(doc.create_element "head")
+            head << style_tag
+          end
+        end
```
Comment (grosser) [inline lib/premailer/adapter/nokogumbo.rb]: inheritance or calling a shared static method would be nice instead of this copy-pasting

Comment (akzhan) [reply lib/premailer/adapter/nokogumbo.rb]: @grosser You are right. my current work have nothing shared with Ruby, so Repo free for your  thoughts.

Comment (lnhrdt) [reply lib/premailer/adapter/nokogumbo.rb]: @grosser where would you like that shared static method to live? I saw the opportunity to share the code but I didn't see any existing examples of code reuse between the adapters apart from the `RgbToHex` module.

File: lib/premailer/adapter/nokogiri_fast.rb (hunk 1 above)
Comment (lnhrdt) [reply lib/premailer/adapter/nokogiri_fast.rb]: good catch

**Comment by grosser** [comment]
don't know/care :D
maybe a new module or something that looks like a base-class ... not sure

File: lib/premailer/adapter/nokogumbo.rb [hunk 3]
```diff
@@ -217,10 +224,11 @@ def load_html(input) # :nodoc:
         # However, we really don't want to hardcode this. ASCII-8BIT should be the default, but not the only option.
         if thing.is_a?(String) and RUBY_VERSION =~ /1.9/
           thing = thing.force_encoding(@options[:input_encoding]).encode!
-          doc = ::Nokogiri::HTML5(thing)
+        end
+        doc = if @options[:html_fragment]
```
Comment (dark-panda) [inline lib/premailer/adapter/nokogumbo.rb]: These lines here appear to be swapped around when compared to the Nokogiri adapter -- `if @options[:html_fragment]` is true, then it should be calling `::Nokogiri::HTML5.fragment(thing)`, not the other way around. This causes an undefined method exception on `#create_element` elsewhere in the code.

Comment (lnhrdt) [reply lib/premailer/adapter/nokogumbo.rb]: Good catch dark-panda. I believe this PR was accepted and released. I'm no longer using Premailer so I don't have time to come back to this. Perhaps @grosser or @akzhan can take a look?

**Pull request merged**